### PR TITLE
in-place reshape ops

### LINKED ISSUE #13998: expand_dims() makes copy instead of simply reshaping
Hi,

Not sure if this is expected behavior or a bug, but I noticed somewhat surprisingly that all `expand_dims()` calls in some code we were writing were making full copies of the underlying array we were expanding. In our case, the only reason to expand was to be able to broadcast arithmetic operations correctly, and it is unacceptable to make a copy. Our current work around is to use reshape directly.

See example code which shows my point:
```
import mxnet as mx

a = mx.nd.array([0])

reshaped = a.reshape(1,1)
reshaped[0,0] = 5
print(a) # prints 5, because reshaped is a wrapper over a                                                                                                                 

expanded = a.expand_dims(1)
expanded[0,0] = 10
print(a) # prints 5, because expanded is a *copy* of a 
```

I personally feel like expand_dims() should have the same behavior as reshape() here, but just be a more fluent way of reshaping an array in a particular way. Therefore I feel like it's a bug for it to make a copy.

However I don't have the full context, and maybe there's a reason expand_dims needs to make a copy? Please let me know--and if not I can look into making a PR to get rid of spurious copy.

Thanks,
Stephen

### PR DESCRIPTION
**Pull request by szha**
## Description ##
make NDArray fluent methods for expand_dims/flatten/squeeze in-place reshapes.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] make NDArray fluent methods for expand_dims/flatten/squeeze in-place reshapes.

## Comments ##
- resolves #13998

### CODE REVIEW

**Comment by ZhennanQin** [comment]
Have you tried this with mkldnn enabled? E.g reshape the output of mkldnn convolution?

**Comment by szha** [comment]
@ZhennanQin I didn't test that specifically based on the assumption that the backend implementation should not cause existing APIs such as `NDArray.reshape` or `NDArray.shape` to break.

**Comment by stephenrawls** [comment]
@szha -- Thanks for putting in this patch!

I have a couple questions about the operator `expand_dims`, which I understand you are actually changing the python code to *not* use that operator, but I think still has relevance.

(1) I originally discovered this because I was using the C API to call MXImperativeInvoke() using the expand_dims operator, and I noticed it was causing a copy. This fix only effects the Python version when operating on ndarrays, not users of the expand_dims operator.

I see that there is a path in the operator that checks if it is an in-place operation. Presumably it uses that path if I pass an output array that is the same NDArrayHandle as the input array? But what if I still need the original input array handle, and I want to create a new output array handle with the expanded dim but still not make a copy?

(2) In the issue I created you commented that: "For symbol (and thus the hybridized version), since in-place identity is possible it should not matter".

Can you talk a little more about that? I assume you mean that in this case:
```
x_expanded = x.expand_dims(1)
y = x_expanded + foo
```
The engine can figure out that x is not needed again, and can thus turn the expand_dims(1) into an in-place operation that doesn't make a copy?

I'm not very familiar with how this part of the code works, so what happens if you had code that looked like this?
```
x_expanded = x.expand_dims(1)
y = x_expanded + foo
z = 2 * x
```
i.e. the code still makes a reference to the original x, and thus presumably the engine can't decide to use the in-place version of expand_dims in that case, right? So I guess my question is -- Does the ability for the Syblolic / hybridized engine to elide the copy depend on the code not referencing the un-expanded version of the array after calling expand_dims()? If so, it seems like there will still be some use cases where an unexpected copy is happening.

**Comment by szha** [comment]
@stephenrawls

> This fix only effects the Python version when operating on ndarrays, not users of the expand_dims operator.

Right. The reshape operators always return a copy. This is because in imperative mode there is no complete computation graph to be analyzed and see whether a copy is needed or not.

> I see that there is a path in the operator that checks if it is an in-place operation. Presumably it uses that path if I pass an output array that is the same NDArrayHandle as the input array?

Right, it is done through `output=original_array` argument to the operators. It does not work for autograd (i.e. Gluon's training mode) as autograd doesn't support in-place operation.

> But what if I still need the original input array handle, and I want to create a new output array handle with the expanded dim but still not make a copy?

As long as it's not used in the context of autograd, you can save the original handle before in-place operation. Though the output of in-place op shares the space with the input, it's still a separate handle.

> For symbol (and thus the hybridized version), since in-place identity is possible it should not matter.

To elaborate, in symbolic mode, the backend can see the complete computation graph, and as a result it can analyze the graph (through node coloring) and figure out whether nodes can share the same space. `FInplaceIdentity` is the graph node attribute that expresses whether this should be allowed in order for compiler to plan memory. For details see [3rdparty/tvm/nnvm/src/pass/plan_memory.cc](https://github.com/dmlc/tvm/blob/0f053c82a747b4dcdf49570ec87c17e0067b7439/nnvm/src/pass/plan_memory.cc).

Note that such memory plan happens before the engine/scheduler work happens, and it happens for both symbolic executor and hybridized Gluon HybridBlock.

**Comment by stephenrawls** [comment]
@szha -- Thanks for the detailed explanation, that helps a lot.

So just for clarity, I see that in the operator definition for expand_dims, it sets the FInPlaceIdentity attribute to true:
https://github.com/apache/incubator-mxnet/blob/45d1a1e6ff9c58cfc75f72651c1bf671ac7f1885/src/operator/tensor/matrix_op.cc#L400-L419

I didn't see anything in the plan_memory.cc file that depended on autograd / training.

So should I take it that the code is smart enough to not allocate new memory for the expand_dims operator in symbolic mode, i.e. not to make a copy, even during training when using autograd?

If so, then I think I understand it, and I see what you mean about only needing to fix this for the Python ndarray use case.

Thanks again!
Stephen

**Comment by vandanavk** [comment]
@mxnet-label-bot add [pr-awaiting-review, NDArray]

**Comment by szha** [comment]
@stephenrawls actually, autograd comes into play in pure imperative mode. MXNet's autograd is trace based and does not support in-place operation such as `input = mx.nd.expand_dims(input, ..., output=input)`. You're actually right in that it doesn't help the case where you write it in imperative mode in C++. In order to make use of the graph pass for saving memory, you need to use Symbol API to express the model and use symbolic execution (or its counter part for Gluon, which is CachedOp).

**Comment by TaoLv** [comment]
Does it mean that with these changes python API will have different behavior compared with other frontend languages?

**Comment by szha** [comment]
@TaoLv Other language bindings don't seem to have the fluent methods for reshaping NDArray.

**Comment by szha** [comment]
The current implementation would break backward compatibility for in-place update.

**Comment by szha** [comment]
I plan on adding an `inplace` argument to these fluent methods to allow users to specify whether in-place operation should be used.

**Comment by ankkhedia** [comment]
@reminisce Could you please help review the PR?

**Comment by szha** [comment]
I haven't finished this PR yet as I haven't got around to add the in-place option.

**Comment by ankkhedia** [comment]
@mxnet-label-bot add [pr-work-in-progress]

**Comment by ankkhedia** [comment]
@mxnet-label-bot remove [pr-awaiting-review]

File: python/mxnet/ndarray/ndarray.py [hunk 1]
```diff
@@ -1701,13 +1765,41 @@ def softmin(self, *args, **kwargs):
...
+            return op.squeeze(self, axis=axis)
+        else:
+            new_shape = list(self.shape)
+            if isinstance(axis, int):
+                axis = [axis]
+            if axis:
+                assert len(axis) == len(set(axis)), \
+                    "axis {} contains duplicate which is not allowed.".format(axis)
+                for i in axis:
+                    assert 0 <= i < len(new_shape), "axis {} is out of range.".format(i)
```
Comment (reminisce) [inline python/mxnet/ndarray/ndarray.py]: Negative axes should be supported.

**Comment by szha** [comment]
@reminisce good catch. I was relying on existing tests before adding the inplace option but forgot to add tests for them afterwards.

**Comment by szha** [comment]
Added tests

**Comment by szha** [comment]
@wkcn good point on symbol and HybridBlock. I'm not sure if a warning should be added as it can get very verbose.

**Comment by szha** [comment]
Nonetheless it should not throw an error when using inplace with hybridize. Let me see how to best deal with it.

**Comment by karan6181** [comment]
@szha Thank you for your contributions! Is this PR work in progress?

**Comment by abhinavs95** [comment]
@szha is this PR ready to merge?

**Comment by piyushghai** [comment]
@szha Is this good to merge ?

**Comment by szha** [comment]
Thanks for all the reviews. I plan to add a dummy option in the symbol version of the reshape ops so that it won't throw error when people use this option in the context of Gluon

**Comment by roywei** [comment]
@mxnet-label-bot add[pr-work-in-progress]

**Comment by roywei** [comment]
@mxnet-label-bot remove [pr-awaiting-response]

**Comment by karan6181** [comment]
@szha Did you get a chance to add a dummy option in the symbol version? Thanks!

**Comment by piyushghai** [comment]
@szha @wkcn This PR is ready to be merged I guess ? 

@mxnet-label-bot Update [NDArray, pr-awaiting-merge]

**Comment by szha** [comment]
@piyushghai thanks for pinging. I wanted to hold onto this for the 1.5 release code freeze.

**Comment by abhinavs95** [comment]
@szha is this good to go now?

**Pull request merged**